kernel/ipi.cc

```diff
@@ -91,52 +91,76 @@ on_ipicall()
   }
 }
 
+/* Logic for implementing pause_other_cpus_and_call
+
+ The core that calls pause_other_cpus_and_call changes the
+ state from Ready to Pausing, Paused to Unpausing, and
+ Unpaused to Ready.
+
+ Of the remaining cores, the last one to pause/unpause will
+ change the state from Pausing to Paused and Unpausing to
+ Unpaused respectively.
+ */
 std::atomic<int> paused_cpu_counter {0};
-enum pause_state { Ready, Pausing, Paused, Unpausing, Unpaused };
-std::atomic<enum pause_state> pause_state {Ready};
-enum pause_state READY = Ready;
+enum pause_state_t { Ready, Pausing, Paused, Unpausing, Unpaused };
+std::atomic<pause_state_t> pause_state {Ready};
 
 // Use IPI to pause other cores until update is complete.
 void
 pause_other_cpus(void)
 {
-  while (!atomic_compare_exchange_strong(&pause_state, &READY, Pausing)); // wait for any previously paused CPUs to resume
+  // wait for any previously paused CPUs to resume
+  pause_state_t READY = Ready;
+  while (!pause_state.compare_exchange_weak(READY, Pausing)) {
+    // atomic::compare_exchange_weak modifies the expected variable
+    // in the failure case, so we need to change it back
+    READY = Ready;
+  }
+
   pushcli();
-  for (int i = 0; i < ncpu; ++i) {
+  for (cpuid_t i = 0; i < ncpu; ++i) {
     if (i != myid())
       lapic->send_pause(&cpus[i]);
   }
-  while (pause_state != Paused); // wait for other CPUs to actually pause
+
+  // wait for other CPUs to actually pause
+  while (pause_state != Paused) nop_pause();
 }
 
 void
-increment_paused_cpu_counter(void)
+resume_other_cpus(void)
 {
-  paused_cpu_counter++;
-  if (paused_cpu_counter == ncpu - 1)
-    pause_state = Paused;
+  pause_state = Unpausing;
+  while (pause_state != Unpaused) nop_pause();
+  popcli();
+  pause_state = Ready;
 }
 
-bool
-is_paused(void)
+void
+pause_other_cpus_and_call(void (*fn)(void))
 {
-  return pause_state == Pausing || pause_state == Paused;
+  if (fn != NULL) {
```

Comment:
It doesn't matter much, but I would probably assert fn is non-null rather than silently do nothing if it is.